### ISSUE TITLE
Decide on validation strategy

### ISSUE THREAD

**Issue by bvaughn** [forms-js/forms-js#16]
This ticket could also read, "_How frequently should we validate?_"

As an example, by default formFor will validate each field on-change and the entire form on-submit. Forms JS could do the same or we could try to optimize and only validate each field once (unless it's changed).

For example, a form consisting of 2 fields: username and password. This form is pre-filled with data. If the user changes the `username` attribute Forms JS will immediately validate the new value. If the user then submits the form, Forms JS _could_ validate _both attributes_ again before allowing the submit- or it could validate only the `password` attribute since `username` has already been validated.

Obviously the 2nd approach is more work to implement but could potentially be worth the effort if a form had many asynchronous custom validators.

What do you all think?

**Comment by kentcdodds** [comment]
I say validate the entire form anytime the model has changed. Reasons:
1. JavaScript is fast - this will not be any application's bottleneck
2. The value of one field can change the validity of another (password and confirm password)
3. Simplifies the API. It's easier to describe to people how it works and easier for people to grok.

**Comment by kentcdodds** [comment]
s/validate the entire field/validate the entire form/

**Comment by bvaughn** [comment]
Validate the entire _form_ anytime a _field_ has changed?

I think I'm leaning towards this option as well, because it's easier and maybe YAGNI.

You raise a good point about one field's value affecting another. There are strategies we could take for that, but maybe the least surprising would be to do a full revalidation.

We'll need to track a field's `pristine` state to avoid showing validation errors before the user has touched a field.

PS. You can edit comments using the little pencil icon. ;)

**Comment by mchapman** [comment]
Couple of considerations:
- JS will not be the bottleneck on all validations - some async validations will involve expensive and unpredictable round trips
- What about an initial validation of the form?  Validation rules change over time, and data read back from a database may no longer satisfy the current validation rules

**Comment by kentcdodds** [comment]
About async validators. Currently with angular's validation pipeline, they have two separate pipelines for validation. The reason for this is so they can avoid running expensive async validation when the cheap sync validation fails. I think we should have a mechanism for people to label validators as expensive which would cause them to be run after the cheap ones (I do not think we should have two different pipelines).

I'm not sure I understand your second consideration

**Comment by mchapman** [comment]
We already discussed only running expensive validations after cheap ones have passed (it was explicit in the [Drive doc](https://docs.google.com/document/d/1cghvrZXMbLzwlG181AuBXzhCR3fmeY06jR17OnID1wA/edit#) and is still implicit) but you could still end up running expensive validations on unchanged fields.

An example of the second point:  a field has validation of enum:['male','female','other']; a user inputs 'other' and persists the data; later the validation changes to that 'male','female','custom' (as per Facebook, but more commonly in my case because I change my mind lots).  When the record is displayed should the field be invalid, even though it is pristine?

**Comment by kentcdodds** [comment]
Ah, understood. Good point about expensive validations running after cheap ones have passed... That's reason enough for me to suggest we only run validations after the field's value has changed, not the entire model. But definitely expose an api where developers can run the validators themselves (like angular's `$validate`). This api would cover the second scenario as well.

**Comment by bvaughn** [comment]
> What about an initial validation of the form?

Using formFor as an example, the controller exposes a method allowing users to manually validate at any point. (More on this below.) This covers the case of forms that want to show error messages immediately. I think it's more common that my users have wanted to suppress messages until the user has either touched a field (no longer _pristine_) or tried to submit the form. This way we're not screaming at users to change something that they were probably going to change anyway.

> Good point about expensive validations running after cheap ones have passed... That's reason enough for me to suggest we only run validations after the field's value has changed

Agreed. Although it doesn't explicitly answer my original question which is: should we do one more full-pass validation before submitting the form?

> But definitely expose an api where developers can run the validators themselves

Yes. This has definitely been a repeated request for formFor and it is why I expect `validateField`, `validateFields` and `validateAll` methods on the form directive controller.

**Comment by bvaughn** [comment]
> We already discussed only running expensive validations after cheap ones have passed (it was explicit in the Drive doc and is still implicit) but you could still end up running expensive validations on unchanged fields.

@mchapman I just re-scanned the doc and I don't see any mention of this. I don't recall discussing it either. Am I spacing? I thought we would always run all of the validators when considering a field (rather than the short-circuiting behavior you're describing). Then our adapters could choose to show all errors or just one at a time. (The nice thing about showing all errors is that users don't end up feeling like they're playing a guessing game- _this field is required! ooh and it has to be 6 characters long! ooh and it also has to have a number! ooh and a capital letter!_ etc.)

I could totally be convinced that this _isn't_ the right way to go. In fact, formFor does the opposite. It starts with the lightweight validations first (e.g. required) and short-circuits on the first failing promise. The async, custom validations (the heaviest ones) are run last and so only get executed if all other validations pass.

**Comment by mchapman** [comment]
I thought it had come up when we were talking about key order in objects vs
arrays, but I may have been thinking it to myself. My thinking was that it
would be sensible to specify the maximum number of failing tests, and the
order of performing them to reach that number quickest.  I agree that
getting loads of errors one at a time is irritating, but I suspect there is
a maximum number of errors it makes sense to report at once, both in terms
of nice presentation and how much a user can be expected to take in.  One
thing that is certain is however flexible it is, a user will want something
that isn't supported!

**Comment by kentcdodds** [comment]
I want to leave the decision of how many errors to display up to the developer. But I do believe that we shouldn't run expensive validators if cheap validators have already failed. I think that's why a distinction between expensive and cheap is useful. I believe that an api to for developers to specify a validator as expensive would be valuable. But I recommend against setting up some complicated priority system. Let's try to make this as simple as possible.

**Comment by bvaughn** [comment]
I'm not sure I'm sold on the concept of specifying cheap vs expensive. I'm tempted to think it will bloat our otherwise lightweight syntax. Currently your validations describe your data constraints. Where would we tack this concept of priority onto? It's a kind of meta data that (I think) doesn't have a clear place within the schemas we've been talking about so far. (I could be convinced to change my mind. This is only an initial opinion.)

We may be prematurely optimizing with this talk of prioritization and short-circuiting. I think we might okay if we just:
- Validation an individual field any time it changes
- Validate all fields any time the form is submitted

I agree with Kent's assessment of leaving the decision of how many errors to display up to the developers (or adapters). Running all validations gives them this choice. Running only until the first failure would take the choice away from them.

**Comment by davidlgj** [comment]
I'm not a fan of  specifying cheap vs expensive either, but what we could do is to separate it into synchronous and asynchronous validations. I guess this would only affect custom validator functions? As Kent said javascript is fast enough, so the synchronous will always be fast. And asynchronous _can_ take a long time. Just throwing it out there even though I'm not sure we need to split them :)

Usually its only forms with fields that depend on each other that can suffer from that we don't validate them all all the time. But they are not that common after all. How about a "group" option? Each field in the validation schema could have an optional "group" that is a string and names a group. Any time a field is validated that belongs to a group the entire group is validated. This way we can support it when needed but we're not stuck doing a lot superfluous validation.

Regarding of if we should stop at first error or keep going, I agree with Brian and Kent that its up to the view layer to decide which validations to display. But that means they all should have a code as well as a message. I want to be able to decide that only the "Required" error message should be shown if present in the list of errors without resorting to comparing strings.

**Comment by kentcdodds** [comment]
I want all validators to be able to return promises whether they're actually hitting a server or not. I don't want two pipelines.

I believe that the "expensive" validators will be used less frequently than the cheap ones. Therefor I think that making it a tiny bit harder to mark them as expensive is not a big deal. This is what I think we could do:

``` javascript
function cheap(value) {
  return value === 'foo';
}

function cheapPromise(value) {
  return new Promise(function(resolve, reject) {
    if (value === 'foo') {
      resolve();
    } else {
      reject();
    }
  });
}

function expensivePromise(value) {
  return doSomethingAsync(value);
  });
}
expensivePromise.expensive = true; // <-- that's the difference
```

This would mean that it is only executed AFTER all the other non expensive ones have passed.

I also considered the idea of actually introducing a `priority` instead, but limit it from 0 to 5 so it doesn't turn into a z-index nightmare... Thoughts?

**Comment by bvaughn** [comment]
> I want all validators to be able to return promises whether they're actually hitting a server or not. I don't want two pipelines.

This is the approach I've taken. Everything is a Promise. Custom validations _might_ be asynchronous Promises.

> This would mean that it is only executed AFTER all the other non expensive ones have passed.

This means we won't be able to give the view all error messages, which means users might resolve all of the errors they're presented with- only to be shown new errors. I don't know that I think this is a big deal, just a side effect we should explicitly acknowledge.

As long as we're just returning a composite Promise I think it's fairly straight-forward to reason about. As we start introducing priority and deferred execution of validations it makes me wish we'd integrate with a library like my [Task Runner](http://www.taskrunnerjs.com/).

**Comment by kentcdodds** [comment]
I think I misunderstood what task runner was initially. If you feel like it
would make development easier/worth the extra dependency, then that's
awesome and I say let's do it.

When trying to picture the scenario you're describing, I don't see a
problem. Let's say we're at a sign up screen that requires an email
address. There are three validators on it: required, email, and unique.
Before they even fill in the form, the input label/placeholder/description
would show that it's a required email field (and maybe that it's required
to be unique). If they skip the field, we just show validation that
indicates its required. If they fill in a non-email, we show them it must
be an email. Even if that invalid email is unique, the user doesn't care
because it wouldn't work anyway.

Does that make sense... Phones are hard to describe things on in detail...
- Kent C. Dodds

(Sent from my mobile device, please forgive typos or brevity)

**Comment by bvaughn** [comment]
I'm very hesitant to add external dependencies, but thanks. We'll see.

I think I'm starting to come around about not returning _all_ validation errors and instead just returning the _first_ error we encounter. The benefits of that are:
- It precludes the whole issue of _expensive_ validations because we can just run custom validations last.
- It takes away view flexibility but it probably doesn't _really_ negatively impact the user.
- Anecdotally, it's what formFor does anyway and no one has complained about it to date.

**Comment by kentcdodds** [comment]
I like having the power to control which of the sync validation errors I
show to the user with angular now... I feel like we'd be making a mistake
to remove that flexibility. I think the approach taken by angular is a good
one. Though I think it was a mistake to assume that async validators were
expensive. I think the developer should have the power to determine which
validators are expensive.
- Kent C. Dodds

(Sent from my mobile device, please forgive typos or brevity)

**Comment by davidlgj** [comment]
What are the arguments for returning a promise from the validating functions? I can only think of one and that is that its nice and clean to have one interface for all validation on a field. 

But I think that shifts complexity over to the view adaptor instead. Which is forced too treat every validation as it could be asynchronous when in fact most of the time its not.  In angular 1.x that would mean you couldn't use the built in `ngModel.$validators` for instance, not that we can't work around it. 

The only validation that I can come up with that is async is validation that hits the server, i.e. looking up if username is taken etc. And that means a custom function right? Sure its a mistake assuming async validators are slow, but if they do something over the network they _can_ be slow.

It might be worth it shifting it over to the adaptor, but since the validator is the part that all share then IMHO its better that it has a bit more complexity and the adaptors can be smaller. 

It would also be easy to split the "validateField" function into three variants, one that does all validating and returns a promise, one that only validates synchronous and one that validates asynchronous so the view adaptor can choose which to work with.  

But since we started prototyping with the asynchronous version why not follow that through and see where it takes us? I might be wrong about the shifting of complexity :)

**Comment by mohsen1** [comment]
I like the initial approach where validators could return boolean, string(error message) or a promise that resolves or rejects to an error message. From a user perspective that would be very easy to work with. In many cases users don't even have an async validators so they don't need to know about promises.

**Comment by bvaughn** [comment]
> What are the arguments for returning a promise from the validating functions...
> 
> I like the initial approach where validators could return boolean, string(error message) or a promise...

I think things are getting muddled. Forms JS base library currently defines a set of built-in validations. I wrote the built-in validations to return Promises because I like consistency. End-users don't need to even be aware of these validators though, or the fact that they are promised-based.

Users can write custom validations that synchronously return truthy/falsy values _or_ promises. There is a built-in validator that knows how to handle both types of user-supplied returns here: https://github.com/forms-js/forms-js/blob/master/source/validation/validators/custom-validator.ts. (**Edit**: @mohsen1 also wanted us to support string return values, so I've just added that.)

How do the views/adapters interact with these validations? That's totally up for discussion, but I'm playing around with some ideas (using lessons learned from formFor) here: https://github.com/forms-js/forms-js/blob/master/source/attribute-metadata.ts

My idea is that an individual input/field could tell an attribute to validate itself and then either listen to the returned Promise _or_ just bind directly to an `errorMessages` object.

**Comment by davidlgj** [comment]
@bvaughn yes I was only thinking about the view/adaptor scenario, not what the user supplies as a validation function. 

I'll do a basic and naive angular 1.x view adaptor to feel it out a bit.

**Comment by bvaughn** [comment]
I know some folks aren't around over the weekend but if it seems preferable, I'm happy to hop on a Hangout this afternoon and chat a bit about this.

My hope here is for the views/adapters to do _very little_. I'd like all of the logic for validation to live in the shared libs library.

**Comment by bvaughn** [comment]
For the Angular 1.x world, here's how I was initially thinking through things:
- A field (directive) would register itself with the Forms JS form and be given (as a return value) an [`AttributeMetadata`](https://github.com/forms-js/forms-js/blob/master/source/attribute-metadata.ts) object.
- This object could be attached to the `$scope` so that the directive can bind to a few of its attributes: `disabled`, `errorMessages`, `required`.
- The directive would then do something like `<input ng-model="username">` and `$scope.$watch('username')`\- and whenever the watch is triggered the directive would call the `AttributeMetadata.validate()`. (No need to use or care about Promises, just fire-and-forget.)
- Once the validation completes (synchronously or asynchronously) the `errorMessages` property of `AttributeMetadata` would be updated and any scope-bindings would be triggered automatically.

This way the Angular touch points are light and there's not much logic to them. The validation all lives inside of Forms JS as a series of Promises, which are easy to composite and to test.

We could use a similar approach in the web-components realm backed by `Object.observe` instead of scope-watch.

PS. If the touch points end up being this light, I think I may even migrate formFor over to Forms JS once we're done.

**Comment by davidlgj** [comment]
Yeah that's basically how i thought about it as well, except I didn't have the AttributeMetadata object in mind :)

Right now I have some simple js code up that generates the directives from the a view (https://github.com/forms-js/angular-1x/tree/simple-generation) and I was hoping to transfer that to typescript and add a directive that does validation tonight or tomorrow. 

When you say this afternoon, could you be more specific?

**Comment by davidlgj** [comment]
...and by tonight I mean after I put the kids to bed in two or three hours :)

**Comment by bvaughn** [comment]
> Yeah that's basically how i thought about it as well, except I didn't have the AttributeMetadata object in mind :)

I think `AttributeMetadata` will make a great bridge between Forms JS and Angular/React/web-components/whatever. It gives adapters something to bind/observe/watch and it hides a lot of the validation complexity.

> add a directive that does validation tonight or tomorrow.

What do you mean by this? :)

If you want to Hangout, I can pretty much make any time work. I'll have to get lunch sometime within the next 2 hours or so but otherwise my schedule is flexible ;)

**Comment by davidlgj** [comment]
So basically in the code I have right now I have one `forms-js` directive that loops over the view and generates a directive depending on "inputType" (as a side note I don't like that name, but "type" is confusing and I don't have a better idea). 

So we end up with something like this: 

``` html
<forms-js view="view" validation="validation">
  <div fjs-text key="username"></div>
  <div fjs-text key="email"></div>
</forms-js>
```

My thought was that the generated directives could also be used "manually".
So then that "$compiles" into something like this:

``` html
<forms-js view="view" validation="validation" model="model">
  <div fjs-text key="username">
     <label>Username</label>
     <input type="text" ng-model="model.username">
  </div>
  <div fjs-text key="email">
      <label>Email</label>
     <input type="text" ng-model="model.username">
  </div>
</forms-js>
```

This is what I have. So the validating directive would go on the actual input (or whatever has the ng-model) like this:
`<input type="text" ng-model="model.username" fjs-validate="username">`

That directive could require the `ngModelController` and add itself to the `$asyncValidators` or as you said just watch the value. It could also use the `$setValidity` to invalidate the field etc

Does that make any sense?

It sometimes takes a while for me to put my kids to bed so I don't dare promise a specific time, but I'm going to try to end their minecraft session now... Go get lunch as usual and we'll just have to see if we can get the timezones to line up :)

**Comment by bvaughn** [comment]
Interesting. Okay, yeah. If our schedules line up after your kids are in bed and I've eaten- it would be helpful to chat a bit.

I'm currently mulling over how a form-submit should happen in a framework-agnostic way so that would be helpful to chat about also. I'm feeling pretty good about the current validation but I'm not sure about submission.

**Comment by mchapman** [comment]
I will be available as well (but probably not contributing much as I am making really slow progress on Angular 2 and TS)

**Comment by davidlgj** [comment]
Great! I'll post a comment when I'm back :)

**Comment by davidlgj** [comment]
Ok. back.

**Comment by davidlgj** [comment]
@bvaughn @mchapman up for a hangout? Otherwise I'm gonna need to crash into bed soon :)

**Comment by bvaughn** [comment]
I will be able to Hangout in 10 minutes. If you're not still up that's okay. I got held up at lunch. Sorry

**Comment by bvaughn** [comment]
Closing this since we've settled on a basic validation strategy.

There was still a question of serial vs parallel validation, and a secondary question of validation priority. I feel we should follow up with these via their own, specific issues though since this is already a pretty long-running thread. I'm personally against the notion of priority but I'm fairly neutral about parallel vs serial. (That is to say that my default stance is...what we have is good enough, so someone else may want to file additional tickets if you disagree.)
